include/vikunja/mem/iterator/PermutationIterator.hpp

```diff
@@ -0,0 +1,274 @@
+#pragma once
+
+#include <alpaka/alpaka.hpp>
+
+#include <iterator>
+
+// check for spaceship support
+#if defined __has_include
+#    if __has_include(<version>)
+#        include <version>
+#    endif
+#    if __has_include(<compare>)
+#        include <compare>
+#        if defined(__cpp_lib_three_way_comparison) && defined(__cpp_impl_three_way_comparison)
+#            define USESPACESHIP
+#        endif
+#    endif
+#endif
+
+#if __has_cpp_attribute(nodiscard)
+#    define NODISCARD [[nodiscard]]
+#else
+#    define NODISCARD
+#endif
+
+namespace vikunja
+{
+    namespace mem
+    {
+        namespace iterator
+        {
+            template<typename ElementIterator, typename IndexIterator, typename IdxType = int64_t>
+            class PermutationIterator
+            {
+            public:
+                using difference_type = int64_t;
+                using value_type = typename std::iterator_traits<ElementIterator>::value_type;
+                using pointer = typename std::iterator_traits<ElementIterator>::pointer;
+                using reference = typename std::iterator_traits<ElementIterator>::reference;
+                using iterator_category = std::random_access_iterator_tag;
+
+                //-----------------------------------------------------------------------------
+                //! Constructor.
+                //!
+                //! \param data Data.
+                ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE
+                PermutationIterator(const ElementIterator& _elIter, const IndexIterator& _idxIter, const IdxType& _idx = static_cast<IdxType>(0))
+                    : elIter(_elIter)
+                    , idxIter(_idxIter)
+                    , idx(_idx)
+                {
+                }
+                
+                /**
+                 * @brief Dereference operator to receive the stored value
+                 */
+                NODISCARD ALPAKA_FN_INLINE reference operator*() const
+                {
+                    return *(elIter + *idxIter);
```

Comment:
The access should be `elIter[idxIter[idx]]]`. Don't forget to check the size of `idxIter`.